tokio/src/sync/mpsc/block.rs

```diff
@@ -28,7 +29,7 @@ pub(crate) struct Block<T> {
     /// Array containing values pushed into the block. Values are stored in a
     /// continuous array in order to improve cache line behavior when reading.
     /// The values must be manually dropped.
-    values: Values<T>,
+    values: Box<Values<T>>,
```

Comment:
This solution adds an extra heap allocation around `values`. I would strongly prefer if we didn't do that.

Reply:
Note that Tokio has this macro: https://github.com/tokio-rs/tokio/blob/master/tokio/src/macros/addr_of.rs

Reply:
The other way around this is something like this:

```rust
#[repr(C)]
struct Block<T> {
    header: BlockHeader<T>,
    values: Values<T>,
}

struct BlockHeader<T> { ... }

impl<T> Block<T> {
    fn new() -> Box<Self> {
         let mut this = /* create an uninitialized Box */;
         (this as *mut BlockHeader<T>).write(/* the starting header*/);
         Values::initialize(this.offset(size_of::<BlockHeader<T>>());
         this
    }
}
```

However, this solution seemed relatively unidiomatic to me. If it's fine with you to do gradual initialization like this then I can do it.

Reply:
Putting the extra fields in a sub-struct is also fine with me.